crates/e2e/tests/e2e/protocol_fee.rs

```diff
@@ -16,334 +21,411 @@ use {
+        .await
+        .unwrap();
 
-async fn surplus_fee_buy_order_test(web3: Web3) {
-    let fee_policy = FeePolicyKind::Surplus {
-        factor: 0.3,
-        max_volume_factor: 0.9,
-    };
-    let protocol_fee = ProtocolFee {
-        policy: fee_policy,
-        policy_order_class: FeePolicyOrderClass::Market,
+    tracing::info!("Waiting for orders metadata update.");
+    let metadata_updated = || async {
+        onchain.mint_block().await;
+        let order_a = services
+            .get_order(&market_price_improvement_uid)
+            .await
+            .unwrap();
+        let order_b = services.get_order(&limit_surplus_order_uid).await.unwrap();
+        let order_c = services.get_order(&partner_fee_order_uid).await.unwrap();
+        !order_a.metadata.executed_surplus_fee.is_zero()
+            && !order_b.metadata.executed_surplus_fee.is_zero()
+            && !order_c.metadata.executed_surplus_fee.is_zero()
     };
-    // Without protocol fee:
-    // Expected execution is 5040413426236634210 GNO for 5000000000000000000 DAI,
-    // with executed_surplus_fee = 167058994203399 GNO
-    //
-    // With protocol fee:
-    // surplus in sell token = 10000000000000000000 - 5040413426236634210 =
-    // 4959586573763365790
-    //
-    // protocol fee in sell token = 0.3*4959586573763365790 = 1487875972129009737
-    //
-    // expected executed_surplus_fee is 167058994203399 + 1487875972129009737 =
-    // 1488043031123213136
-    //
-    // Settlement contract balance after execution = executed_surplus_fee GNO
-    execute_test(
-        web3.clone(),
-        vec![ProtocolFeesConfig(vec![protocol_fee])],
-        OrderKind::Buy,
-        None,
-        1488043031123213136u128.into(),
-        1488043031123213136u128.into(),
-    )
-    .await;
+    wait_for_condition(TIMEOUT, metadata_updated).await.unwrap();
+
+    let market_price_improvement_order = services
+        .get_order(&market_price_improvement_uid)
+        .await
+        .unwrap();
+    let market_executed_surplus_fee_in_buy_token =
+        market_price_improvement_order.metadata.executed_surplus_fee
+            * market_quote_after.buy_amount
+            / market_quote_after.sell_amount;
+    let market_quote_diff = market_quote_after
+        .buy_amount
+        .saturating_sub(market_quote_before.quote.buy_amount);
+    assert!(market_executed_surplus_fee_in_buy_token >= market_quote_diff * 3 / 10);
+
+    let partner_fee_order = services.get_order(&partner_fee_order_uid).await.unwrap();
+    let partner_fee_executed_surplus_fee_in_buy_token =
+        partner_fee_order.metadata.executed_surplus_fee * partner_fee_quote_after.buy_amount
+            / partner_fee_quote_after.sell_amount;
+    let partner_fee_quote_diff = partner_fee_quote_after
+        .buy_amount
+        .saturating_sub(partner_fee_order.data.buy_amount);
+    assert!(partner_fee_executed_surplus_fee_in_buy_token >= partner_fee_quote_diff * 2 / 100);
```

Comment:
Still the subject of investigation. The difference could possibly be explained by a fee from the solver scaled in the buy token: https://github.com/cowprotocol/services/blob/a50a2e669c279160d6ac1da6d0144bfded66411e/crates/solvers/src/domain/solution.rs#L168
While for limit and market orders, it almost works(the remaining difference is `2640619288`), for the partner fee, it doesn't, meaning either `partner_fee_quote_diff` is computed wrongly or the assumption is wrong.